app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -111,89 +101,30 @@ public void afterTextChanged(final Editable s) {
+        _lineNumbersDrawer.setDefaultPaddingLeft(getPaddingLeft());
     }
 
     @Override
     public boolean onPreDraw() {
-        _paint.setTextSize(getTextSize());
+        _lineNumbersDrawer.setTextSize(getTextSize());
         return super.onPreDraw();
     }
 
     @Override
     protected void onDraw(Canvas canvas) {
         super.onDraw(canvas);
 
-        // If line numbers can be drawn
         if (_numEnabled) {
-            drawLineNumbers(canvas);
-        } else if (getPaddingLeft() != _defaultPaddingLeft) {
-            // Reset padding without line numbers fence
-            setPadding(_defaultPaddingLeft, getPaddingTop(), getPaddingRight(), getPaddingBottom());
-        }
-    }
-
-    private void drawLineNumbers(final Canvas canvas) {
-        if (!getLocalVisibleRect(_numRect)) {
-            return;
-        }
-
-        final CharSequence text = getText();
-        final Layout layout = getLayout();
-        if (text == null || layout == null) {
-            return;
-        }
-
-        // Current visible area
-        final int height = _hlRect.height();
-        final int top = _hlRect.top - height;
-        final int bottom = _hlRect.bottom + height;
-
-        final int offsetY = getPaddingTop();
-        final int maxLayoutLine = layout.getLineCount();
-
-        // We make a single pass through the text to determine
-        // 1. y positions and line numbers we want to draw
-        // 2. max line number (to gauge gutter width)
-        _numYPositions.clear();
-        int startNumber = 1, maxNumber = 1;
-        for (int i = 0; i < maxLayoutLine; i++) {
-            final int start = layout.getLineStart(i);
-            if (start == 0 || text.charAt(start - 1) == '\n') {
-                final int numY = layout.getLineBaseline(i);
-                if (numY < top) {
-                    startNumber++;
-                } else if (numY < bottom) {
-                    _numYPositions.add(numY);
-                }
-                maxNumber++;
-            }
-        }
-
-        // Draw the gutter
-        final int width = Math.round(_paint.measureText(String.valueOf(maxNumber)));
-        final int numX = LINE_NUMBERS_PADDING_LEFT + width;
-        final int lineX = numX + LINE_NUMBERS_PADDING_RIGHT;
-        final int padding = lineX + 10;
-        if (padding != getPaddingLeft()) {
-            setPadding(padding, getPaddingTop(), getPaddingRight(), getPaddingBottom());
-        }
-        _paint.setColor(Color.LTGRAY);
-        canvas.drawLine(lineX, top, lineX, bottom, _paint);
-
-        _paint.setColor(Color.GRAY);
-        for (int i = 0; i < _numYPositions.size(); i++) {
-            final String number = String.valueOf(startNumber + i);
-            final float numY = _numYPositions.get(i) + offsetY;
-            canvas.drawText(number, numX, numY, _paint);
+            _lineNumbersDrawer.draw(canvas);
+        } else {
+            _lineNumbersDrawer.suspend();
```

Comment:
Instead of calling this here I think calling this in setLineNumbersEnabled(false) is sufficient.

Reply:
Fine, I will improve it.